### ISSUE TITLE
Fix for issue #129 makes freezing time in threads impossible

### ISSUE THREAD

**Issue by nickwilliams-eventbrite** [spulec/freezegun#307]
I'm trying to test the interaction of several threads. In doing so, I need to freeze time, assert some things, tick, assert some more things, tick, etc. However, the fix for #129 appears to have made this impossible to do.

The problem is that I'm using `time.time()` in the code that I'm testing, and that code runs within multiple threads. As a result, the `threading` module appears in the stacktrace for my code that uses `time.time()`. So when `_should_use_real_time` inspects the stacktrace, it sees the `threading` frame and returns `True`, causing my code to get the real time. I had to do this to fix it:

```python
def patched_freeze_time():
    f = freezegun.freeze_time()
    f.ignore = tuple(set(f.ignore) - {'threading'})
    return f

with patched_freeze_time() as frozen_time:
    # run tests
```

I understand the usefulness of ignoring threading by default, and I don't want to regress #129, so I suggest a new `freeze_time` argument-with-default `ignore_threading=True`, and if it set to `False` then the `threading` module will not be included in `ignore`.

At alternative fix, if you feel that it's truly necessary to always use real time in calls within the `threading` module itself, would be to hard-code exclude `threading` from the step where you patch modules' imports, and then remove `threading` from the default ignores. In doing this, code calling `time.time()` (for example) within a running thread would still be affected by the frozen time, but `time.time()` calls within the `threading` module itself would not and would use real time.

**Comment by spulec** [comment]
Yeah, I agree that we should provide a way for the user to more granularly control which modules are ignored or not. I would prefer a more generic solution though that works for all of the ignored modules.

Maybe something like a `add_default_ignores=True` parameter in the API. Then, we only append to ignores if that is True. If a user passes false, then they are responsible for passing in everything they want ignored in the `ignore` parameter.

**Comment by boxed** [comment]
I think we might need to think of more advanced/complex rules for looking at the stack. Not just saying "It's threading!" but looking at why it's in that module.

**Comment by nickwilliams-eventbrite** [comment]
Actually, I think I have a better idea than _either_ `ignore_threading=True` _or_ `add_default_ignores=True`. Currently, there's an argument `ignore=None` argument, and if it's specified, the value of it is added to the default ignores (which are hard-coded appended in the function). The other problem with that is that it's using lists, not sets, so if the default ignores and the input ignores include duplicates, you have duplicates. So I would refactor like this:

```python
DEFAULT_IGNORES = frozenset({'nose.plugins', 'six.moves', 'threading', etc....})

def freeze_time(......., ignore=(), ....., do_not_ignore=()):
    ...
    ignore = ignore or ()  # in case existing code is passing in `None`, which the default used to be
    ignore = DEFAULT_IGNORES + set(ignore) - set(do_not_ignore)
    
    return _freeze_time(..., ignore, ...)
```

**Comment by spulec** [comment]
Agree that a more advanced system would be ideal! If there are ideas, I think that would be great. Otherwise, I like the idea of allowing users to modify the default list.

**Comment by westover** [comment]
Following

**Comment by sloria** [comment]
I've been using the workaround in the [OP](https://github.com/spulec/freezegun/issues/307#issue-469349090), but it's no longer working as of https://github.com/spulec/freezegun/pull/369 , which is preventing us from upgrading to versions newer than 1.0.0

**Comment by boxed** [comment]
@sloria You might want to consider switching to time-machine. It's maintained and more robust.
